source/connection_monitor.c

```diff
@@ -0,0 +1,200 @@
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: channel throughput - %" PRIu64 " bytes per second",
+        (void *)channel,
+        bytes_per_second);
+
+    if (bytes_per_second >= impl->options.minimum_throughput_bytes_per_second) {
+        return;
+    }
+
+    /*
+     * We failed the throughput check.  Restore and increment the failure count and then check if the failure threshold
+     * has been crossed.
+     */
+    impl->consecutive_throughput_failures = current_failure_count + 1;
+    AWS_LOGF_INFO(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: Channel low throughput warning.  Currently %u consecutive failures",
+        (void *)channel,
+        impl->consecutive_throughput_failures);
+
+    if (impl->consecutive_throughput_failures < impl->options.minimum_throughput_failure_threshold_in_seconds) {
+        return;
+    }
+
+    AWS_LOGF_INFO(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: Channel low throughput threshold hit (< %" PRIu64 " bytes per second for %u seconds).  Shutting down.",
+        (void *)channel,
+        impl->options.minimum_throughput_bytes_per_second,
+        impl->options.minimum_throughput_failure_threshold_in_seconds);
+
+    aws_channel_shutdown(channel, AWS_ERROR_HTTP_CHANNEL_THROUGHPUT_FAILURE);
+}
+
+static void s_destroy(struct aws_crt_statistics_handler *handler) {
+    if (handler == NULL) {
+        return;
+    }
+
+    aws_mem_release(handler->allocator, handler);
+}
+
+static uint64_t s_get_report_interval_ms(struct aws_crt_statistics_handler *handler) {
+    (void)handler;
+
+    return 1000;
+}
+
+static struct aws_crt_statistics_handler_vtable s_http_connection_monitor_vtable = {
+    .process_statistics = s_process_statistics,
+    .destroy = s_destroy,
+    .get_report_interval_ms = s_get_report_interval_ms,
+};
+
+struct aws_crt_statistics_handler *aws_crt_statistics_handler_new_http_connection_monitor(
+    struct aws_allocator *allocator,
+    struct aws_http_connection_monitoring_options *options) {
+    struct aws_crt_statistics_handler *handler = NULL;
+    struct aws_statistics_handler_http_connection_monitor_impl *impl = NULL;
+
+    if (!aws_mem_acquire_many(
+            allocator,
+            2,
+            &handler,
+            sizeof(struct aws_crt_statistics_handler),
+            &impl,
+            sizeof(struct aws_statistics_handler_http_connection_monitor_impl))) {
+        return NULL;
+    }
+
+    AWS_ZERO_STRUCT(*handler);
+    AWS_ZERO_STRUCT(*impl);
+    impl->options = *options;
+
+    handler->vtable = &s_http_connection_monitor_vtable;
+    handler->allocator = allocator;
+    handler->impl = impl;
+
+    return handler;
+}
+
+bool aws_http_connection_monitoring_options_is_valid(const struct aws_http_connection_monitoring_options *options) {
+    if (options == NULL) {
+        return false;
+    }
+
+    return options->minimum_throughput_failure_threshold_in_seconds > 1 &&
+           options->minimum_throughput_bytes_per_second > 0;
```

Comment:
The comments warn that 1 might result in flaky behavior, but here you outright refuse to do monitoring if they use it

Reply:
Modified the name and meaning a bit.